presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,283 @@
+        {
+            List<PlanNode> sources = joinGraph.getSources();
+            List<PlanNode> leftSources = partitioning.stream()
+                    .map(sources::get)
+                    .collect(toImmutableList());
+            List<PlanNode> rightSources = ImmutableList.copyOf(Sets.difference(ImmutableSet.copyOf(sources), ImmutableSet.copyOf(leftSources)));
+            Set<Symbol> leftSymbols = leftSources.stream()
+                    .flatMap(node -> node.getOutputSymbols().stream())
+                    .collect(toImmutableSet());
+
+            InnerJoinPredicateUtils.InnerJoinPushDownResult pushDownResult = sortPredicatesForInnerJoin(leftSymbols, ImmutableList.of(joinGraph.getFilter()), ImmutableList.of());
+
+            Set<Symbol> joinSymbols = ImmutableSet.<Symbol>builder()
+                    .addAll(joinGraph.getOutputSymbols())
+                    .addAll(DependencyExtractor.extractUnique(pushDownResult.getJoinPredicate()))
+                    .build();
+            PlanNode left = getJoinSource(
+                    idAllocator,
+                    leftSources,
+                    pushDownResult.getLeftPredicate(),
+                    joinSymbols.stream().filter(leftSymbols::contains).collect(toImmutableList()));
+            PlanNode right = getJoinSource(
+                    idAllocator,
+                    rightSources,
+                    pushDownResult.getRightPredicate(),
+                    joinSymbols.stream()
+                            .filter(symbol -> !leftSymbols.contains(symbol))
+                            .collect(toImmutableList()));
+
+            List<Expression> joinPredicates = extractConjuncts(pushDownResult.getJoinPredicate());
+            List<JoinNode.EquiJoinClause> joinConditions = joinPredicates.stream()
+                    .filter(JoinEnumerator::isJoinEqualityCondition)
+                    .map(predicate -> toEquiJoinClause((ComparisonExpression) predicate, leftSymbols))
+                    .collect(toImmutableList());
+            List<Expression> joinFilters = joinPredicates.stream()
+                    .filter(predicate -> !isJoinEqualityCondition(predicate))
+                    .collect(toImmutableList());
+
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> inputSymbols = ImmutableList.<Symbol>builder()
+                    .addAll(left.getOutputSymbols())
+                    .addAll(right.getOutputSymbols())
+                    .build();
+            List<Symbol> outputSymbols = inputSymbols.stream()
+                    .filter(joinGraph.getOutputSymbols()::contains)
+                    .collect(toImmutableList());
+            return new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    flipJoinCriteriaIfNecessary(joinConditions, left),
```

Comment:
why not to instantiate `joinConditions` in proper order in `toEquiJoinClause`? This way you would avoid additional `flipJoinCriteriaIfNecessary`